---
title: "Genetic Association Study with lazyGas"
author: 
  -name: "Tomoyuki Furuta"
  affiliation: Institute of Plant Science and Resources, Okayama University, Okayama, Japan
  email: f.tomoyuki@okayama-u.ac.jp
date: "Jun 26, 2024"
package: lazyGas
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Genetic Association Study with lazyGa}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction
This vignette demonstrates the process of conducting a genetic association study using the `lazyGas` package, along with the `GBScleanR` package for data cleaning. We will go through the steps of loading input files, preprocessing data, performing association analysis, and visualizing results. 

## Prerequisites
Ensure you have the following packages installed and loaded:
```{r}
library(lazyGas)
library(GBScleanR)
library(ggplot2)
```

## Create sample genotype and phenotype data
```{r}
library(GenovisR)
library(SNPRelate)
library(SeqArray)
library(Biostrings)

nb <- readDNAStringSet("~/hdd3/genomeData/rice/cultivar_sativa/os_nb_rapdb/IRGSP-1.0_genome.fasta")

df <- simGenovis(n_sample = 100, n_chr = 12, chr_len = width(nb) * 1e-6)
temp_dir <- tempdir()
temp_gds <- tempfile("temp", temp_dir, ".gds")
snpgdsCreateGeno(gds.fn = temp_gds,
                 genmat = df$genotype,
                 sample.id = df$sample_info$id,
                 snp.id = seq_along(df$marker_info$id),
                 snp.rs.id = seq_along(df$marker_info$id),
                 snp.chromosome = df$marker_info$chr,
                 snp.position = df$marker_info$pos,
                 snp.allele = paste(df$marker_info$ref_allele,
                                    df$marker_info$alt_allele, sep = "/"),
                 snpfirstdim = FALSE, compress.annotation = "LZMA_RA",
                 compress.geno = "LZMA_RA",
                 other.vars = list(HAP = df$haplotype,
                                   EDS = df$dosage))

sample_gds <- tempfile("sample", temp_dir, ".gds")
seqSNP2GDS(gds.fn = temp_gds,
           out.fn = sample_gds)

gds <- openfn.gds(sample_gds, readonly = FALSE)

addfolder.gdsn(node = index.gdsn(gds, "annotation/format"), name = "HAP")
addfolder.gdsn(node = index.gdsn(gds, "annotation/format"), name = "EDS")
add.gdsn(index.gdsn(gds, "annotation/format/HAP"), name = "data",
         val = df$haplotype, storage = "bit2", compress = "LZMA_RA")
add.gdsn(index.gdsn(gds, "annotation/format/EDS"), name = "data",
         val = df$dosage, storage = "bit2", compress = "LZMA_RA")
closefn.gds(gds)
unlink(temp_gds)

qtl <- sample(x = df$marker_info$id, size = 3)
qtl_index <- which(df$marker_info$id %in% qtl)
df$marker_info[qtl_index, ]
qtl_ds <- df$dosage[, qtl_index]

e <- rnorm(n = nrow(qtl_ds), mean = 0, sd = 1)
q1_add <- qtl_ds[, 1] * 3
q1_dom <- as.numeric(qtl_ds[, 1] == 1) * 3
q2_add <- qtl_ds[, 2] * 2
q2_dom <- as.numeric(qtl_ds[, 2] == 1) * 0
q3_add <- qtl_ds[, 3] * 1
q3_dom <- as.numeric(qtl_ds[, 3] == 1) * 2

st <- 1^2 + 3^2 + 2^2 + 1^2
3^2/st
2^2/st
1^2/st

sample_pheno <- tempfile("pheno", temp_dir, ".csv")
pheno <- q1_add + q1_dom + q2_add + q2_dom + q3_add + q3_dom + e
pheno <- data.frame(id = df$sample_info$id, pheno = pheno)
write.csv(pheno, sample_pheno, row.names = FALSE)
```


## Loading Input GDS File and Building a LazyGas Object
Load the input GDS file and build a `LazyGas` class object:
```{r}
gds_fn <- system.file("extdata", "sample.gds", package = "lazyGas")
lg <- buildLazyGas(gds_fn = gds_fn, load_filter = TRUE, overwrite = FALSE)
```

If you have genotype information that are not stored in a GDS file, you can build 
a LazyGas class object by supplying a named list including the necessary 
information to the `create_gds` argument of the `buildLazyGas()` function.
```{r}
temp_dir <- tempdir()

# Now we assume genotype and marker data were provided as a matrix or 
# data.frame that have been loaded from CSV files, although here we use data 
# obtained from the GDS file for ease of handling.
genotype <- getGenotype(object = lg, node = "raw")
dosage <- getGenotype(object = lg, node = "dosage")
haplotype <- getHaplotype(object = lg)
snp.chromosome <- getChromosome(object = lg)
snp.position <- getPosition(object = lg)
snp.allele <- getAllele(object = lg)
sample.id <- getSamID(object = lg)
snp.id <- getMarID(object = lg)

# The following values can be NULL: snp.id, snp.rs.id, and snp.allele
# Either of the following values should be specified at least: genotype, 
# haplotype, or dosage.
create_gds <-  list(genotype = genotype,
                    sample.id = sample.id,
                    snp.id = snp.id,
                    snp.rs.id = NULL,
                    snp.chromosome = snp.chromosome,
                    snp.position = snp.position,
                    snp.allele = snp.allele,
                    haplotype = haplotype,
                    dosage = dosage)
sample_gds <- tempfile("sample", temp_dir, ".gds")
lg <- buildLazyGas(gds_fn = sample_gds, 
                   create_gds = create_gds)
```


## Loading and Preprocessing Phenotype Data
Load the phenotype data:
```{r}
pheno_fn <- system.file("extdata", "pheno.csv", package = "lazyGas")
pheno <- read.csv(file = pheno_fn)
```

Assign phenotype data to the LazyGas object:
```{r}
lg <- assignPheno(object = lg,
                  pheno = pheno,
                  rename = "Fruit weight")
```

## Visualizing Phenotype Data
Generate and save plots for phenotype data:
```{r}
pheno <- getPheno(object = lg)
for(i in seq_along(pheno$pheno_names)){
  p <- plotPheno(object = lg, pheno = i, xlab = pheno$pheno_names[i])
  print(p)
}
```


## Conducting Association Analysis

### Prepare a function to make a model matrix.
Confirming structure of genotype data to be used for the regression analysis
```{r}
g <- getGenoPerMarker(object = lg, geno_format = "genotype")
print(g)
```

```{r}
g <- getGenoPerMarker(object = lg, geno_format = "haplotype")
print(g)
```

```{r}
g <- getGenoPerMarker(object = lg, geno_format = "dosage")
print(g)
```

As shown above, genotype and haplotype data for each marker is a matrix with two
rows if you have diploid samples, whereas dosage data is a vector.
You need to prepare a function to convert those data into a model matrix for the 
regression analysis.

Although you can manually define a function for the data conversion, 
the makeConvFun() function helps you create a data conversion function. 

Manually define a function to convert genotype data to a model matrix:
```{r}
# Conversion function for dosage data assuming additive and dominant effects
conv_fun <- function(g){
  add <- g
  dom <- as.numeric(g == 1)
  out <- data.frame(add = add, dom = dom)
  return(out)
}

# For the model matrix above, the formula can be the following.
formula <- "add + dom"

# Conversion function for haplotype data assuming additive and interaction effects
conv_fun <- function(g){
  hap1 <- colSums(g == 1)
  hap2 <- colSums(g == 2)
  hap1_hap2 <- as.numeric(hap1 == 1 & hap2 == 1)
  out <- data.frame(hap2 = hap2, hap1_hap2 = hap1_hap2)
  return(out)
}
formula <- "hap2 + hap1_hap2"

```

Define a conversion function by makeConvFun():
```{r}
conv_fun <- makeConvFun(geno_format = "dosage", n_levels = 3)
```
```{r}
formula <- "add + dom"
```


The current implementation of the makeConvFun() function does not support 
to assume interaction effects of haplotypes. If necessary, please define a 
conversion function manually as shown in the previous code chunk. When you have
four haplotypes, the function definition can be the following.
```{r eval=FALSE}
# Do not run.
conv_fun <- function(g){
  hap1 <- colSums(g == 1)
  hap2 <- colSums(g == 2)
  hap3 <- colSums(g == 3)
  hap4 <- colSums(g == 4)
  hap1_hap2 <- as.numeric(hap1 == 1 & hap2 == 1)
  hap1_hap3 <- as.numeric(hap1 == 1 & hap3 == 1)
  hap1_hap4 <- as.numeric(hap1 == 1 & hap4 == 1)
  hap2_hap3 <- as.numeric(hap2 == 1 & hap3 == 1)
  hap2_hap4 <- as.numeric(hap2 == 1 & hap4 == 1)
  hap3_hap4 <- as.numeric(hap3 == 1 & hap4 == 1)
  out <- data.frame(hap2 = hap2, hap3 = hap3, hap4 = hap4, 
                    hap1_hap2 = hap1_hap2, hap1_hap3 = hap1_hap3,
                    hap1_hap4 = hap1_hap4, hap2_hap3 = hap2_hap3,
                    hap2_hap4 = hap2_hap4, hap3_hap4 = hap3_hap4)
  return(out)
}
formula <- paste0("hap2 + hap3 + hap4 + hap1_hap2 + hap1_hap3",
                  " + hap1_hap4 + hap2_hap3 + hap2_hap4 + hap3_hap4")
```


Run the association scan:
```{r}
scanAssoc(object = lg,
          formula = formula,
          conv_fun = conv_fun,
          geno_format = "dosage")
```

For GWAS, set `method = "mlm"`.
```{r}
scanAssoc(object = lg,
          geno_format = "genotype",
          method = "mlm")
```

If you conducted association tests using another tool and have p values,
you can assign those p values to the input lazyGas class object.
```{r}
# As an example, extract p values from the lazyGas object and assign them to the object again.
scan <- lazyData(object = lg, dataset = "scan", pheno = "Fruit weight")
lg <- assignPvalues(object = lg, pheno_name = "Fruit weight", p_values = scan$P.model)
```

## Visualizing Results
Generate Manhattan plots for association results:
```{r}
pheno <- getPheno(object = lg)
for(i in seq_along(pheno$pheno_names)){
  p <- plotManhattan(object = lg, pheno = i)
  p <- p + labs(title = pheno$pheno_names[i])
  print(p)
}

scan <- lazyData(object = lg, dataset = "scan", pheno = pheno$pheno_names[1])
scan[which.max(scan$negLog10P), ]
```

Call and visualize peak blocks:
```{r}
callPeakBlock(object = lg, signif = 0.05, threshold = 0.8)

for(i in seq_along(pheno$pheno_names)){
  p <- plotPeaks(object = lg, pheno = i, recalc = FALSE)
  p <- p + labs(title = pheno$pheno_names[i])
  print(p)
}
```

## Recalculating and Visualizing Peaks
Recalculate and visualize peaks:
```{r}
recalcAssoc(object = lg, n_threads = 10, refine_position = FALSE)

for(i in seq_along(pheno$pheno_names)){
  p <- plotPeaks(object = lg, pheno = i, recalc = TRUE)
  p <- p + labs(title = pheno$pheno_names[i])
  print(p)
}
```

## Visualizing Haplotype and Phenotype Data
Generate plots for haplotype and phenotype data:
```{r}
for(i in seq_along(pheno$pheno_names)){
  out <- haploPlot(object = lg, pheno = i, recalc = FALSE)
  print(out)
}
```


```{r}
for(i in seq_along(pheno$pheno_names)){
  out <- haploPlot(object = lg, pheno = i, recalc = TRUE)
  print(out)
}
```

## List up candidate genes
You can list up candidate genes responsisble for the detected associations as 
following. 
```{r}
# GFF file 
# gff_fn <- "path/to/gff/file.gff"
gff_fn <- "~/01_wd/genomeData/rice/cultivar_sativa/nb_combined/version_2023/gff/nb_combined_all_annotated.gff"

# SNPEff output VCF file
snpeff_fn <- "/home/ftom/01_wd/yamamoto/shufle/snpeff/founder_wgs_on_bv_genome_snpeff.vcf.gz"

# Any CSV format file for gene annotations
# The given matrix stored in the CSV file must contain a column named "GeneID"
# that should match the gene IDs in the GFF file.
# ann_fn <- "path/to/annotation/file.csv" 
source("~/01_wd/genomeData/util_scripts/util08_orgAnnotation.R")
gff <- rtracklayer::import.gff(gff_fn)
ann <- rtracklayer::mcols(gff)
ann <- ann[ann$type %in% c("gene", "mRNA", "transcript"), ]
ann <- subset(ann, select = c(gene_id, RAP_Note, MSU_Note, MapMan_DESCRIPTION, Oryzabase_Trait_Ontology, RAPDB_geneSymbol, Oryzabase_geneSymbol, InterPro_Description))
ann <- as.data.frame(ann)
ann <- orgAnnotation(id = ann$gene_id, ann = ann)
names(ann)[1] <- "Gene_ID"
ann <- ann[, -2]
# listCandidate(object = lg, gff_fn = gff_fn, snpeff_fn = snpeff_fn, 
#               ann = ann, recalc = TRUE, numerize_chr = TRUE)
```


## Extracting Data
Extract data for further analysis:
```{r}
pheno <- getPheno(object = lg)
sapply(pheno, head)
```


```{r}
scan <- lazyData(object = lg, dataset = "scan", pheno = pheno$pheno_names[1])
head(scan)
```


```{r}
peakcall <- lazyData(object = lg, dataset = "peakcall", pheno = pheno$pheno_names[1])
head(peakcall)
```


```{r}
recalc <- lazyData(object = lg, dataset = "recalc", pheno = pheno$pheno_names[1])
head(recalc)
```

```{r}
groups <- lazyData(object = lg, dataset = "groups", pheno = pheno$pheno_names[1])
head(groups)
```

```{r}
candidate <- lazyData(object = lg, dataset = "candidate", pheno = pheno$pheno_names[1])
head(candidate)
```

```{r}
candidate <- lazyData(object = lg, dataset = "snpeff", pheno = pheno$pheno_names[1])
head(candidate)
```
```{r}
makeInteractiveSummary(object = lg, pheno = pheno$pheno_names[1], out_fn = "sample.html")
```

