---
title: "Genetic Association Study with LazyGas"
author: 
  -name: "Tomoyuki Furuta"
  affiliation: Institute of Plant Science and Resources, Okayama University, Okayama, Japan
  email: f.tomoyuki@okayama-u.ac.jp
date: "Jun 26, 2024"
package: LazyGas
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Genetic Association Study with LazyGas}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction
This vignette demonstrates the process of conducting a genetic association study using the `lazyGas` package, along with the `GBScleanR` package for data cleaning. We will go through the steps of loading input files, preprocessing data, performing association analysis, and visualizing results. 

## Prerequisites
Ensure you have the following packages installed and loaded:
```{r}
library(lazyGas)
library(GBScleanR)
library(ggplot2)
```

## Loading Input GDS File and Building a LazyGas Object
Load the input GDS file and build a `LazyGas` class object:
```{r}
gds_fn <- system.file("extdata", "sample.gds", package = "lazyGas")
lg <- buildLazyGas(gds_fn = gds_fn, load_filter = TRUE, overwrite = FALSE)
```

If you have genotype information that are not stored in a GDS file, you can build 
a LazyGas class object by supplying a named list including the necessary 
information to the `create_gds` argument of the `buildLazyGas()` function.
```{r}
temp_dir <- tempdir()

# Now we assume genotype and marker data were provided as a matrix or 
# data.frame that have been loaded from CSV files.
genotype <- getGenotype(object = lg, node = "raw")
dosage <- getGenotype(object = lg, node = "dosage")
haplotype <- getHaplotype(object = lg)
snp.chromosome <- getChromosome(object = lg)
snp.position <- getPosition(object = lg)
snp.allele <- getAllele(object = lg)
sample.id <- getSamID(object = lg)
snp.id <- getMarID(object = lg)

# The following values can be NULL: snp.id, snp.rs.id, and snp.allele
# Either of the following values should be specified at least: genotype, 
# haplotype, or dosage.
create_gds <-  list(genotype = genotype,
                    sample.id = sample.id,
                    snp.id = snp.id,
                    snp.rs.id = NULL,
                    snp.chromosome = snp.chromosome,
                    snp.position = snp.position,
                    snp.allele = snp.allele,
                    haplotype = haplotype,
                    dosage = dosage)
lg <- buildLazyGas(gds_fn = file.path(temp_dir, "sample.gds"), 
                   create_gds = create_gds)
```


## Loading and Preprocessing Phenotype Data
Load the phenotype data:
```{r}
pheno_fn <- system.file("extdata", "pheno.csv", package = "lazyGas")
pheno <- read.csv(file = pheno_fn)
```

Assign phenotype data to the LazyGas object:
```{r}
lg <- assignPheno(object = lg,
                  pheno = pheno,
                  rename = "Fruit weight")
```

## Visualizing Phenotype Data
Generate and save plots for phenotype data:
```{r}
pheno <- getPheno(object = lg)
for(i in seq_along(pheno$pheno_names)){
  p <- plotPheno(object = lg, pheno = i, xlab = pheno$pheno_names[i])
  print(p)
}
```


## Conducting Association Analysis

### Prepare a function to make a model matrix.
Confirming structure of genotype data to be used for the regression analysis
```{r}
g <- getGenoPerMarker(object = lg, geno_format = "genotype")
print(g)
```

```{r}
g <- getGenoPerMarker(object = lg, geno_format = "haplotype")
print(g)
```

```{r}
g <- getGenoPerMarker(object = lg, geno_format = "dosage")
print(g)
```

As shown above, genotype and haplotype data for each marker is a matrix with two
rows if you have diploid samples, whereas dosage data is a vector.
You need to prepare a function to convert those data into a model matrix for the 
regression analysis.

Although you can manually define a function for the data conversion, 
the makeConvFun() function helps you create a data conversion function. 

Manually define a function to convert genotype data to a model matrix:
```{r}
# Conversion function for dosage data assuming additive and dominant effects
conv_fun <- function(g){
  add <- g
  dom <- as.numeric(g == 1)
  out <- data.frame(add = add, dom = dom)
  return(out)
}

# For the model matrix above, the formula can be the following.
formula <- "add + dom"

# Conversion function for haplotype data assuming additive and interaction effects
conv_fun <- function(g){
  hap1 <- colSums(g == 1)
  hap2 <- colSums(g == 2)
  hap1_hap2 <- as.numeric(hap1 == 1 & hap2 == 1)
  out <- data.frame(hap2 = hap2, hap1_hap2 = hap1_hap2)
  return(out)
}
formula <- "hap2 + hap1_hap2"

```

Define a conversion function by makeConvFun():
```{r}
conv_fun <- makeConvFun(geno_format = "dosage", n_levels = 3)
```
```{r}
formula <- "add + dom"
```


The current implementation of the makeConvFun() function does not support 
to assume interaction effects of haplotypes. If necessary, please define a 
conversion function manually as shown in the previous code chunk. When you have
four haplotypes, the function definition can be the following.
```{r eval=FALSE}
# Do not run.
conv_fun <- function(g){
  hap1 <- colSums(g == 1)
  hap2 <- colSums(g == 2)
  hap3 <- colSums(g == 3)
  hap4 <- colSums(g == 4)
  hap1_hap2 <- as.numeric(hap1 == 1 & hap2 == 1)
  hap1_hap3 <- as.numeric(hap1 == 1 & hap3 == 1)
  hap1_hap4 <- as.numeric(hap1 == 1 & hap4 == 1)
  hap2_hap3 <- as.numeric(hap2 == 1 & hap3 == 1)
  hap2_hap4 <- as.numeric(hap2 == 1 & hap4 == 1)
  hap3_hap4 <- as.numeric(hap3 == 1 & hap4 == 1)
  out <- data.frame(hap2 = hap2, hap3 = hap3, hap4 = hap4, 
                    hap1_hap2 = hap1_hap2, hap1_hap3 = hap1_hap3,
                    hap1_hap4 = hap1_hap4, hap2_hap3 = hap2_hap3,
                    hap2_hap4 = hap2_hap4, hap3_hap4 = hap3_hap4)
  return(out)
}
formula <- paste0("hap2 + hap3 + hap4 + hap1_hap2 + hap1_hap3",
                  " + hap1_hap4 + hap2_hap3 + hap2_hap4 + hap3_hap4")
```


Run the association scan:
```{r}
scanAssoc(object = lg,
          formula = formula,
          conv_fun = conv_fun,
          geno_format = "dosage",
          kruskal = NULL)
```

## Visualizing Results
Generate Manhattan plots for association results:
```{r}
pheno <- getPheno(object = lg)
for(i in seq_along(pheno$pheno_names)){
  p <- plotManhattan(object = lg, pheno = i)
  p <- p + labs(title = pheno$pheno_names[i])
  print(p)
}

scan <- lazyData(object = lg, dataset = "scan", pheno = pheno$pheno_names[1])
scan[which.max(scan$negLog10P), ]
```

Call and visualize peak blocks:
```{r}
callPeakBlock(object = lg, signif = 0.05, threshold = 0.8)

for(i in seq_along(pheno$pheno_names)){
  p <- plotPeaks(object = lg, pheno = i, recalc = FALSE)
  p <- p + labs(title = pheno$pheno_names[i])
  print(p)
}
```

## Recalculating and Visualizing Peaks
Recalculate and visualize peaks:
```{r}
recalcAssoc(object = lg, signif = 0.05, threshold = 0.8)

for(i in seq_along(pheno$pheno_names)){
  p <- plotPeaks(object = lg, pheno = i, recalc = TRUE)
  p <- p + labs(title = pheno$pheno_names[i])
  print(p)
}
```

## Visualizing Haplotype and Phenotype Data
Generate plots for haplotype and phenotype data:
```{r}
for(i in seq_along(pheno$pheno_names)){
  out <- haploPlot(object = lg, pheno = i, recalc = FALSE)
  print(out)
}

for(i in seq_along(pheno$pheno_names)){
  out <- haploPlot(object = lg, pheno = i, recalc = TRUE)
  print(out)
}
```

## List up candidate genes
You can list up candidate genes responsisble for the detected associations as 
following. 
```{r eval=FALSE}
# GFF file 
gff_fn <- "path/to/gff/file.gff"

# SNPEff output VCF file
snpeff_fn <- "path/to/snpeff/file.vcf"

# Any CSV format file for gene annotations
# The given matrix stored in the CSV file must contain a column named "GeneID"
# that should match the gene IDs in the GFF file.
ann_fn <- "path/to/annotation/file.csv" 

listCandidate(object = lg, gff = gff_fn, snpeff = snpeff_fn, 
              ann = ann_fn, recalc = TRUE)
```


## Extracting Data
Extract data for further analysis:
```{r}
pheno <- getPheno(object = lg)
sapply(pheno, head)
```


```{r}
scan <- lazyData(object = lg, dataset = "scan", pheno = pheno$pheno_names[1])
head(scan)
```


```{r}
peakcall <- lazyData(object = lg, dataset = "peakcall", pheno = pheno$pheno_names[1])
head(peakcall)
```


```{r}
recalc <- lazyData(object = lg, dataset = "recalc", pheno = pheno$pheno_names[1])
head(recalc)
```

```{r eval=FALSE}
candidate <- lazyData(object = lg, dataset = "candidate", pheno = pheno$pheno_names[1])
head(candidate)
```
